libnmstate/nm/checkpoint.py

```diff
@@ -92,21 +47,115 @@ def __exit__(self, exc_type, exc_val, exc_tb):
     def create(self):
         devs = []
         timeout = self._timeout
-        cp_flags = (
-            CHECKPOINT_CREATE_FLAG_DESTROY_ALL |
-            CHECKPOINT_CREATE_FLAG_DELETE_NEW_CONNECTIONS |
-            CHECKPOINT_CREATE_FLAG_DISCONNECT_NEW_DEVICES
-        )
-        cp = self._manager.interface.CheckpointCreate(devs, timeout, cp_flags)
-        logging.debug('Checkpoint %s created for all devices: %s', cp, timeout)
-        self._checkpoint = cp
+        cp_flags = (nmclient.NM.CheckpointCreateFlags.DESTROY_ALL |
+                    nmclient.NM.CheckpointCreateFlags.DELETE_NEW_CONNECTIONS |
+                    nmclient.NM.CheckpointCreateFlags.DISCONNECT_NEW_DEVICES)
+        userdata = self
+
+        self._mainloop.push_action(
+            self._nmclient.checkpoint_create,
+            devs,
+            timeout,
+            cp_flags,
+            self._mainloop.cancellable,
+            _checkpoint_create_callback,
+            userdata)
+
+        if not self._mainloop.run(timeout):
+            raise CheckPointError("Failed to create checkpoint")
 
     def destroy(self):
-        self._manager.interface.CheckpointDestroy(self._checkpoint)
-        logging.debug('Checkpoint %s destroyed', self._checkpoint)
+        if self._checkpoint:
+            userdata = self
+            self._mainloop.push_action(
+                self._nmclient.checkpoint_destroy,
+                self._checkpoint.get_path(),
+                self._mainloop.cancellable,
+                _checkpoint_destroy_callback,
+                userdata)
+            self._checkpoint = None
```

Comment:
Shouldn't you destroy it in `_checkpoint_destroy_callback`?
Same for the rollback.

Reply:
Done.